### PR TITLE
Proc macro updates

### PR DESCRIPTION
**Pull request by birkenfeld**
This is based on #191, #181 and the `use-lisp-fn` branch.

### CODE REVIEW

**Comment by Wilfred** [comment]
Looks like this is just failing due to rustfmt, though I imagine we'll get the OS X failure once rustfmt is happy.

File: rust_src/remacs-macros/function.rs [hunk 1]
```diff
@@ -78,81 +87,29 @@ pub fn get_fn_arg_ident(fn_arg: &syn::FnArg) -> Result<syn::Ident> {
...
                     ArgType::Other => {
-                        return Err("lisp functions should contain only `LispObject`s");
+                        return Err("lisp functions should only have `LispObject` args");
                     },
                 }
             },
             _ => return Err("lisp functions cannot have `self` arguments"),
         }
     }
-
```
Comment (jeandudey) [inline rust_src/remacs-macros/function.rs]: :100:, this looked like a mounstrosity. Thanks for doing this :smile:

File: rust_src/remacs-macros/lisp_attr.rs [hunk 2]
```diff
@@ -1,63 +1,72 @@
...
-use std::collections::HashMap;
 use syn;
 use synom;
 
+use function::{Function, LispFnType};
+
+/// Arguments of the lisp_fn attribute.
 pub struct LispFnArgs {
+    /// Desired Lisp name of the function.
+    /// If not given, derived as the Rust name with "_" -> "-".
```
Comment (jeandudey) [inline rust_src/remacs-macros/lisp_attr.rs]: :+1: Very ingenious

**Comment by birkenfeld** [comment]
Thanks for the reviews! I'll rebase this onto master, then any conflicts should be gone.

**Comment by Wilfred** [comment]
@birkenfeld it's odd that the tests are still failing. I'll take a look. Would you mind pushing your old version of this branch somewhere? I'd like to bisect the various changes before they were squashed.

**Comment by birkenfeld** [comment]
@Wilfred unfortunately I don't have that reflog with me currently. But it's still only Tramp, so your suggestion to disable the tests might be incomplete?

**Comment by birkenfeld** [comment]
Merged via #203.

**Pull request closed without merging**